### DISCUSSION TITLE
how can a client exposes method to the server

### DISCUSSION THREAD

**Discussion by sasidhar-ss** [Kistler-Group/sdbus-cpp#472] [Q&A]
Hi,

I have Clients (multiple) and a server that uses dbus for communication.  Client uses sdbus c++ library, and server uses a different one. The client could send requests to the server and receive a response from the server for the request that it sent. Now I want to expose a method (recv_msg) from the client so that the the server will asynchronously call that method. Every client exposes the same method (recv_msg) , so the server will call the exposed-method using the Client's bus id. I have the below implementation and it didn't work.  someone, please share your thoughts on this.

```
class PubSubClient {
public:
    PubSubClient() {
        // Create proxy object for the object on the server side
        sdbus::ServiceName server_serviceName{"com.example.PubSub"};
	sdbus::ObjectPath server_objectPath{"/com/example/PubSub"};
	connection = sdbus::createBusConnection();
        std::string uniqueName1 = connection->getUniqueName();
        std::cout << "Client1 D-Bus unique name (for server connection): " << uniqueName1 << std::endl;
        proxy_ = sdbus::createProxy(*connection, std::move(server_serviceName), std::move(server_objectPath));
        auto recv_msg = [this](const std::string msg)  -> std::string{
            std::string result = "received";
            std::cout << "Msg received : " << msg << std::endl;
            return result;
        };
        sdbus::ObjectPath client_objectPath{"/com/example/Client1"};
        clientObject_ = sdbus::createObject(*connection, std::move(client_objectPath));
        sdbus::InterfaceName interfaceName{"com.example.Client1"};
        // Register the "CreateTopic" method
        clientObject_->addVTable(sdbus::registerMethod("recv_msg").implementedAs(std::move(recv_msg)))
                               .forInterface(interfaceName);
       std::thread eventLoopThread([&]() {
            connection->enterEventLoop(); // Blocking call
       });
        eventLoopThread.join();
    };
```
    
Best Regards,
sam.

**Comment by sangelovic** [comment]
The additional `connectionToServer_` is a bit confusing there -- where does it come from? I suppose you mean `connection` instance instead. Otherwise I'm not 100% sure off the top of my head, but I think it should work that you have a proxy and an object on the same connection. What exactly does not work? What behavior do you get? Does at least calling a server method from the client work? Does calling `recv_msg` from the server work? Make sure you have [permissions set up](https://github.com/Kistler-Group/sdbus-cpp/blob/master/docs/systemd-dbus-config.md) correctly....

**Reply by sangelovic** [reply]
Btw., wouldn't signals suffice in your case? So a server, when its method is called or whenever it wants to inform clients or given something like to clients, simply emits a signal and clients would react to the signal. This is actually a nice pub-sub implementation, and a simple one, that D-Bus is designed for.

**Reply by sasidhar-ss** [reply]
@sangelovic Thanks for your response. 
Sorry for the typo error. Yes, it is **connection->getUniqueName()**. Calling a server method from the client worked, but when the server calls the client method (recv_msg) using the client ID, the client hasn't received any data and the server prints a dbus timeout error after a few seconds. 
I also tried calling the "recv_msg" method using dbus-send, and there is no message received by the client app .
*  dbus-send --session --dest=:1.13908 /com/example/Client1 com.example.Client1.recv_msg  string:"Hello from dbus-send"
* both the client and the server are not running as a systemd service. Do I still need to setup the permissions?
* Is it possible for the server to emit the signal to a particular client based on the client ID, instead of broadcasting the signal to all the clients?
I am new to D-Bus and the sdbus library; apologies for the fundamental questions.
